controller/internal/controller/consumer_controller.go

```diff
@@ -39,7 +41,20 @@ import (
 // ConsumerReconciler reconciles a Consumer object
 type ConsumerReconciler struct {
 	component.ResourceManager
-	Output component.Output
+	Output   component.Output
+	keyIndex *KeyIndexRegistry // Add a new Key index
+}
+
+type KeyIndexRegistry struct {
+	mu    sync.RWMutex
```

Comment:
We could use sync.Mutext, since only write lock is used. The `sync.Mutex` is faster than the write lock of RWMutex.